### DISCUSSION TITLE
session authentication with next-auth like express-session

### DISCUSSION THREAD

**Discussion by codelover2k** [nextauthjs/next-auth#7405] [Help]
### Question 💬

Hi!

in all my servers I am using express and express-session.
Since I started some pure next.js and tRPC apps I want to use next-auth.

My question is if it is possible to use next-auth like express-session?
express-session allows you crazy customization for your session and session-cookie like the way you want to protect it with a secret to sign it, maxAge and so on.

My flow looks like this with express-session that I want to have in next-auth.
1. I check if the req has a session
2. I get the session and from my database/redis I get the session and its data
3. I use req.session.userId to fetch my database to fetch user informations like the name, its permissions, if the user is locked by the admin or not and many more for every request
4. return the requested data to the user or return a "NOT AUTHENTICATED" respond to the client

is this possible with next-auth?
It looks and feels like you have to use an authentication service like google and that only JWT are supported?

### How to reproduce ☕️

---

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by balazsorban44** [comment]
"Crazy customization" is kind of vague, but we do have a lot of configuration options. Check the docs https://next-auth.js.org/configuration/options

We also support database sessions out of the box, Redis included.

https://authjs.dev/reference/adapter/upstash-redis

You can also, create your custom database adapter.

https://authjs.dev/guides/adapters/creating-a-database-adapter

https://github.com/nextauthjs/next-auth/blob/main/packages/adapter-upstash-redis/src/index.ts

**Reply by codelover2k** [reply]
hi @balazsorban44 and thanks for respond!

So in my express servers, that is what I use with REST or GraphQL.
My cookies are configured like this with **express-session**:
```js
  app.use(
    session({
      name: "COOKIE_ID",
      store: new RedisStore({
        client: redis,
        disableTouch: true,
      }),
      cookie: {
        maxAge: 1000 * 60 * 60 * 24, // 1 day
        httpOnly: true,
        sameSite: "lax",  // reLAXed CSRF - Cross Site Request Forgery
        secure: true, // cookie only works in https
        domain: ".myapp.com", // cookie only works if request comes from this domain
      },
      saveUninitialized: false,  // not every session will be stored, only modified once
      secret: "very complicated string", // this string is used to sign the cookie and protect it from modifications.
      resave: false, // will not save the session in store on every request, only if was modified
    })
  );
```
when I login, I can store data, which is only the id of the user and I put it in `req.session.userId`.

now I can protect my routes like this:
```js
const protected = async (req, res) => {
  // no userId in session means no previous and successful login
  if (!req.session.userId) {
    throw Unauthenticated("you are not logged in and have no rights to see that");
  }

  // get the latest user data. Maybe the admin or the user changed something a second ago
  const userData = await getUserById(req.session.userId);

  // will only be there if login was successful
  if (!userdata) {
    throw Unauthenticated("you are not logged in and have no rights to see that")
  }

  if (userdata.isLocked) {
    throw Locked("you are locked by the admin and can not use this service anymore")
  }

  // set up the user in session, so other routes can use it for authorization and other stuff
  req.session.userData = userData;

 // everything is okay, jump to the next middleware
 next();
};

app.get("/secret-data", protected, (req, res) => { res.send("very important and secret data") });
```
Can you please help me to get something like that in next-auth?

I do not see how I can do that with the session object of next-auth to get the same settings like I get in express-session.
I simple example would be great 😊

**Reply by balazsorban44** [reply]
https://next-auth.js.org/getting-started/example#backend---api-route

**Reply by codelover2k** [reply]
I did see that, but is it normal that next-auth session has so minimal options?

So the main question is how can you configure the session in next-auth and its behavior like express-session allows you?
Is that the minimum that you can configure for a session in next-auth?
```js
export default NextAuth({
     providers: [
       // add your authentication providers
     ],
     session: {
       jwt: false, // or strategy: "database"?
       maxAge: 60 * 60 * 24, // 24 hours
       updateAge: 60 * 10, // everything is in memory, but in 10 minutes touch and update the store
       store: new SessionProvider(redisClient),
       // add any additional options
       /*
       ------------ WHERE ARE THESE OPTIONS ? ------------ :
       - domain
       - sameSite: "strict" OR "lax"
       - httpOnly - to only allow http as a protocol
       - secure flag - to allow https only
       - saveUninstalized - flag to tell next-auth whether to store a session, even if it is not modified
       - resave - flag to tell next-auth to save the session in store on every request or not
       */
     },
   })
```
These are the options that I am missing for my cookie session.
Maybe they exist, but just with another name?

If it is possible, can you help me to set it up and configure it as in my express-session example?
